{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Daisy Algorithm for ImmeDB results\n",
    "\n",
    "1. For a new row of '20230614_AMR in TE_Abricate-report' googlesheet, open the ImmeDB results for the strain that corresponds to the row. \n",
    "2. In the ImmeDB result file, search hits with \"q.start\" value smaller than \"START\" value of Abricate (Column \"C\" in '20230614_AMR in TE_Abricate-report'; mobile genetic element region should start prior to antibiotic resistance gene). \n",
    "3. Among those, search hits with \"q.end\" value larger than \"END\" value of Abricate (Column \"D\" in '20230614_AMR in TE_Abricate-report'; mobile genetic element region should end after the antibiotic resistance gene).\n",
    "4. For a hit that satisfies both 2&3, get \"subject acc.ver\" in ImmeDB results file (ex; NZ_NMTU01000006.1:71946-83411).\n",
    "5. Search the \"subject acc.ver\" in \"Data1_MGE_sequences.fasta\" file to figure out the name of the hit (ex;  IMEs459). \n",
    "    - If there are multiple hits, any ICE elements take priority to document. \n",
    "    - If there's no hit, put 'na' in column O (ImmeDB values) of '20230614_AMR in TE_Abricate-report' googlesheet\n",
    "6. Copy the name of the hit to column O (ImmeDB values) of '20230614_AMR in TE_Abricate-report' googlesheet.\n",
    "7. Copy the corresponding information from the row in the ImmeDB result file (from column B-P) and paste it to column P-AD of '20230614_AMR in TE_Abricate-report' googlesheet. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cloudpathlib import AnyPath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FILE</th>\n",
       "      <th>SEQUENCE</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>GENE</th>\n",
       "      <th>COVERAGE</th>\n",
       "      <th>COVERAGE_MAP</th>\n",
       "      <th>GAPS</th>\n",
       "      <th>%COVERAGE</th>\n",
       "      <th>%IDENTITY</th>\n",
       "      <th>DATABASE</th>\n",
       "      <th>ACCESSION</th>\n",
       "      <th>PRODUCT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>./Inputfasta/SH0001344-00329.fasta</td>\n",
       "      <td>scaffold_Scaffolds_0.cp.600</td>\n",
       "      <td>3402126</td>\n",
       "      <td>3404045</td>\n",
       "      <td>+</td>\n",
       "      <td>tet(O)</td>\n",
       "      <td>1-1920/1920</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>99.9</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>NG_048267.1</td>\n",
       "      <td>tetracycline resistance ribosomal protection p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>./Inputfasta/SH0001365-00245.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>43127</td>\n",
       "      <td>44353</td>\n",
       "      <td>+</td>\n",
       "      <td>mef(A)</td>\n",
       "      <td>1-1227/1227</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>97.64</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>NG_047970.1</td>\n",
       "      <td>macrolide efflux MFS transporter Mef(A)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>./Inputfasta/SH0001372-00039.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>3220448</td>\n",
       "      <td>3222364</td>\n",
       "      <td>-</td>\n",
       "      <td>tet(W)</td>\n",
       "      <td>1-1917/1923</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>99.69</td>\n",
       "      <td>96.19</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>NG_048295.1</td>\n",
       "      <td>tetracycline resistance ribosomal protection p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>./Inputfasta/SH0001373-00076.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>670898</td>\n",
       "      <td>671463</td>\n",
       "      <td>+</td>\n",
       "      <td>catA16</td>\n",
       "      <td>95-660/660</td>\n",
       "      <td>..=============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>85.76</td>\n",
       "      <td>99.29</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>NG_047591.1</td>\n",
       "      <td>type A-16 chloramphenicol O-acetyltransferase ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>./Inputfasta/SH0001404-00077.fasta</td>\n",
       "      <td>scaffold_Scaffolds_0.cp.600</td>\n",
       "      <td>1387439</td>\n",
       "      <td>1389355</td>\n",
       "      <td>-</td>\n",
       "      <td>tet(W)</td>\n",
       "      <td>1-1917/1923</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>99.69</td>\n",
       "      <td>97.91</td>\n",
       "      <td>ncbi</td>\n",
       "      <td>NG_048295.1</td>\n",
       "      <td>tetracycline resistance ribosomal protection p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 FILE                     SEQUENCE    START  \\\n",
       "0  ./Inputfasta/SH0001344-00329.fasta  scaffold_Scaffolds_0.cp.600  3402126   \n",
       "1  ./Inputfasta/SH0001365-00245.fasta                            1    43127   \n",
       "2  ./Inputfasta/SH0001372-00039.fasta                            1  3220448   \n",
       "3  ./Inputfasta/SH0001373-00076.fasta                            1   670898   \n",
       "4  ./Inputfasta/SH0001404-00077.fasta  scaffold_Scaffolds_0.cp.600  1387439   \n",
       "\n",
       "       END STRAND    GENE     COVERAGE     COVERAGE_MAP GAPS %COVERAGE  \\\n",
       "0  3404045      +  tet(O)  1-1920/1920  ===============  0/0       100   \n",
       "1    44353      +  mef(A)  1-1227/1227  ===============  0/0       100   \n",
       "2  3222364      -  tet(W)  1-1917/1923  ===============  0/0     99.69   \n",
       "3   671463      +  catA16   95-660/660  ..=============  0/0     85.76   \n",
       "4  1389355      -  tet(W)  1-1917/1923  ===============  0/0     99.69   \n",
       "\n",
       "  %IDENTITY DATABASE    ACCESSION  \\\n",
       "0      99.9     ncbi  NG_048267.1   \n",
       "1     97.64     ncbi  NG_047970.1   \n",
       "2     96.19     ncbi  NG_048295.1   \n",
       "3     99.29     ncbi  NG_047591.1   \n",
       "4     97.91     ncbi  NG_048295.1   \n",
       "\n",
       "                                             PRODUCT  \n",
       "0  tetracycline resistance ribosomal protection p...  \n",
       "1            macrolide efflux MFS transporter Mef(A)  \n",
       "2  tetracycline resistance ribosomal protection p...  \n",
       "3  type A-16 chloramphenicol O-acetyltransferase ...  \n",
       "4  tetracycline resistance ribosomal protection p...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abricate_file = AnyPath(\"~/Downloads/20230614_AMR-Abricate_report.csv\").expanduser()\n",
    "\n",
    "abricate_df = (\n",
    "    pd.read_csv(\n",
    "        abricate_file,\n",
    "        sep=\",\",\n",
    "        header=0,\n",
    "        index_col=False,\n",
    "        dtype=str,\n",
    "        usecols=set(range(0, 14)),\n",
    "    )\n",
    "    .rename(columns={\"#FILE\": \"FILE\"})\n",
    "    .sort_values(by=[\"FILE\"])\n",
    ")\n",
    "\n",
    "abricate_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "immedb_file = AnyPath(\n",
    "    \"s3://genomics-workflow-core/scratch/sunitj/daisy/immeDB/Data1_MGE_sequences.annotations.csv\"\n",
    ")\n",
    "immedb_df = pd.read_csv(\n",
    "    immedb_file,\n",
    "    sep=\",\",\n",
    "    header=0,\n",
    "    index_col=False,\n",
    "    dtype=str,\n",
    "    names=[\"accession\", \"annotation\"],\n",
    ")\n",
    "\n",
    "immedb_annotations = dict(zip(immedb_df[\"accession\"], immedb_df[\"annotation\"]))\n",
    "# immedb_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['query', 'subject', 'perc_identity', 'alignment_length', 'mismatches',\n",
       "       'gap_opens', 'q_start', 'q_end', 's_start', 's_end', 'evalue',\n",
       "       'bit_score', 'query_length', 'subject_length',\n",
       "       'perc_query_coverage_per_subject', 'subject sci names', 'ImmeDB_values',\n",
       "       'All_immeDB_annotations'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "\n",
    "\n",
    "def read_immedb_result(genome_name: str, seq_name: str, start: int, end: int):\n",
    "    result_columns = [\n",
    "        \"query\",\n",
    "        \"subject\",\n",
    "        \"perc_identity\",\n",
    "        \"alignment_length\",\n",
    "        \"mismatches\",\n",
    "        \"gap_opens\",\n",
    "        \"q_start\",\n",
    "        \"q_end\",\n",
    "        \"s_start\",\n",
    "        \"s_end\",\n",
    "        \"evalue\",\n",
    "        \"bit_score\",\n",
    "        \"query_length\",\n",
    "        \"subject_length\",\n",
    "        \"perc_query_coverage_per_subject\",\n",
    "        \"subject sci names\",\n",
    "    ]\n",
    "    # s3://genomics-workflow-core/Results/Blast/MITI-MCB/SH0001342-00095/immeDB/SH0001342-00095.blastn.tsv\n",
    "    immedb_file = AnyPath(\n",
    "        f\"s3://genomics-workflow-core/Results/Blast/MITI-MCB/{genome_name}/immeDB/{genome_name}.blastn.tsv\"\n",
    "    )\n",
    "    df = (\n",
    "        pd.read_table(\n",
    "            immedb_file,\n",
    "            comment=\"#\",\n",
    "            header=None,\n",
    "            names=result_columns,\n",
    "            dtype={\"query\": str, \"subject\": str},\n",
    "        )\n",
    "        .query(\"(query == @seq_name) and (q_start <= @start) and (q_end >= @end)\")\n",
    "        .assign(ImmeDB_values=lambda x: x[\"subject\"].map(immedb_annotations))\n",
    "    )\n",
    "    # return df\n",
    "\n",
    "    if df.empty:\n",
    "        return [None] * (len(result_columns) + 2)\n",
    "\n",
    "    # add freq of also found categories.\n",
    "    uniq_annot = Counter(df[\"ImmeDB_values\"].to_list())\n",
    "    df_0 = df.iloc[0].copy()\n",
    "    df_0[\"All_immeDB_annotations\"] = uniq_annot\n",
    "\n",
    "    return df_0\n",
    "\n",
    "\n",
    "# read_immedb_result(\"SH0001372-00039\", \"1\", 3220448, 3222364)\n",
    "read_immedb_result(\n",
    "    \"SH0001344-00329\", \"scaffold_Scaffolds_0.cp.600\", 3402126, 3404045\n",
    ").index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_columns = [\n",
    "    \"query\",\n",
    "    \"subject\",\n",
    "    \"perc_identity\",\n",
    "    \"alignment_length\",\n",
    "    \"mismatches\",\n",
    "    \"gap_opens\",\n",
    "    \"q_start\",\n",
    "    \"q_end\",\n",
    "    \"s_start\",\n",
    "    \"s_end\",\n",
    "    \"evalue\",\n",
    "    \"bit_score\",\n",
    "    \"query_length\",\n",
    "    \"subject_length\",\n",
    "    \"perc_query_coverage_per_subject\",\n",
    "    \"subject sci names\",\n",
    "    \"ImmeDB_values\",\n",
    "    \"All_immeDB_annotations\",\n",
    "]\n",
    "abricate_df[result_columns] = abricate_df.apply(\n",
    "    lambda x: read_immedb_result(\n",
    "        AnyPath(x.FILE).stem, x.SEQUENCE, int(x.START), int(x.END)\n",
    "    ),\n",
    "    axis=1,\n",
    "    result_type=\"expand\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FILE</th>\n",
       "      <th>SEQUENCE</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>GENE</th>\n",
       "      <th>COVERAGE</th>\n",
       "      <th>COVERAGE_MAP</th>\n",
       "      <th>GAPS</th>\n",
       "      <th>%COVERAGE</th>\n",
       "      <th>...</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bit_score</th>\n",
       "      <th>query_length</th>\n",
       "      <th>subject_length</th>\n",
       "      <th>perc_query_coverage_per_subject</th>\n",
       "      <th>subject sci names</th>\n",
       "      <th>ImmeDB_values</th>\n",
       "      <th>All_immeDB_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>./Inputfasta/SH0001344-00329.fasta</td>\n",
       "      <td>scaffold_Scaffolds_0.cp.600</td>\n",
       "      <td>3402126</td>\n",
       "      <td>3404045</td>\n",
       "      <td>+</td>\n",
       "      <td>tet(O)</td>\n",
       "      <td>1-1920/1920</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>11001.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>20116.0</td>\n",
       "      <td>3890762.0</td>\n",
       "      <td>11001.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs1</td>\n",
       "      <td>{'IMEs1': 20, 'ICEs78': 1, 'IMEs393': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>./Inputfasta/SH0001365-00245.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>43127</td>\n",
       "      <td>44353</td>\n",
       "      <td>+</td>\n",
       "      <td>mef(A)</td>\n",
       "      <td>1-1227/1227</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>9585.0</td>\n",
       "      <td>8157.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2368.0</td>\n",
       "      <td>2324782.0</td>\n",
       "      <td>11179.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Genomic_islands50</td>\n",
       "      <td>{'Genomic_islands50': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>./Inputfasta/SH0001372-00039.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>3220448</td>\n",
       "      <td>3222364</td>\n",
       "      <td>-</td>\n",
       "      <td>tet(W)</td>\n",
       "      <td>1-1917/1923</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>99.69</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8027.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14277.0</td>\n",
       "      <td>4962077.0</td>\n",
       "      <td>8027.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs42</td>\n",
       "      <td>{'IMEs42': 3, 'IMEs201': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>./Inputfasta/SH0001373-00076.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>670898</td>\n",
       "      <td>671463</td>\n",
       "      <td>+</td>\n",
       "      <td>catA16</td>\n",
       "      <td>95-660/660</td>\n",
       "      <td>..=============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>85.76</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>./Inputfasta/SH0001404-00077.fasta</td>\n",
       "      <td>scaffold_Scaffolds_0.cp.600</td>\n",
       "      <td>1387439</td>\n",
       "      <td>1389355</td>\n",
       "      <td>-</td>\n",
       "      <td>tet(W)</td>\n",
       "      <td>1-1917/1923</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>99.69</td>\n",
       "      <td>...</td>\n",
       "      <td>193.0</td>\n",
       "      <td>2435.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3938.0</td>\n",
       "      <td>2294929.0</td>\n",
       "      <td>8027.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs42</td>\n",
       "      <td>{'IMEs42': 3, 'IMEs201': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>253</th>\n",
       "      <td>./Inputfasta/SH0002528-00199.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>328911</td>\n",
       "      <td>330131</td>\n",
       "      <td>+</td>\n",
       "      <td>tet(40)</td>\n",
       "      <td>1-1221/1221</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>9442.0</td>\n",
       "      <td>10836.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2499.0</td>\n",
       "      <td>3770667.0</td>\n",
       "      <td>10836.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs393</td>\n",
       "      <td>{'IMEs393': 1, 'IMEs460': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>./Inputfasta/SH0002528-00199.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>326851</td>\n",
       "      <td>328770</td>\n",
       "      <td>+</td>\n",
       "      <td>tet(O)</td>\n",
       "      <td>1-1920/1920</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2757.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5092.0</td>\n",
       "      <td>3770667.0</td>\n",
       "      <td>11001.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs1</td>\n",
       "      <td>{'IMEs1': 20, 'ICEs78': 1, 'IMEs393': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>./Inputfasta/SH0002528-00199.fasta</td>\n",
       "      <td>1</td>\n",
       "      <td>2718930</td>\n",
       "      <td>2719796</td>\n",
       "      <td>+</td>\n",
       "      <td>aadE</td>\n",
       "      <td>1-867/867</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>1648.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2863.0</td>\n",
       "      <td>3770667.0</td>\n",
       "      <td>8189.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs35</td>\n",
       "      <td>{'IMEs35': 3, 'IMEs363': 1, 'IMEs369': 1, 'ICE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>./Inputfasta/SH0002531-00147.fasta</td>\n",
       "      <td>scaffold_1.cp.600</td>\n",
       "      <td>839789</td>\n",
       "      <td>841708</td>\n",
       "      <td>+</td>\n",
       "      <td>tet(O)</td>\n",
       "      <td>1-1920/1920</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>10430.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15830.0</td>\n",
       "      <td>3022989.0</td>\n",
       "      <td>10938.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs1</td>\n",
       "      <td>{'IMEs1': 20, 'ICEs78': 1, 'IMEs393': 1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>./Inputfasta/SH0002531-00147.fasta</td>\n",
       "      <td>scaffold_1.cp.600</td>\n",
       "      <td>2253295</td>\n",
       "      <td>2253789</td>\n",
       "      <td>+</td>\n",
       "      <td>dfrF</td>\n",
       "      <td>1-495/495</td>\n",
       "      <td>===============</td>\n",
       "      <td>0/0</td>\n",
       "      <td>100</td>\n",
       "      <td>...</td>\n",
       "      <td>1664.0</td>\n",
       "      <td>997.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1223.0</td>\n",
       "      <td>3022989.0</td>\n",
       "      <td>15726.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IMEs383</td>\n",
       "      <td>{'IMEs383': 1, 'IMEs324': 1, 'IMEs419': 1, 'IM...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>257 rows Ã— 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   FILE                     SEQUENCE    START  \\\n",
       "0    ./Inputfasta/SH0001344-00329.fasta  scaffold_Scaffolds_0.cp.600  3402126   \n",
       "1    ./Inputfasta/SH0001365-00245.fasta                            1    43127   \n",
       "2    ./Inputfasta/SH0001372-00039.fasta                            1  3220448   \n",
       "3    ./Inputfasta/SH0001373-00076.fasta                            1   670898   \n",
       "4    ./Inputfasta/SH0001404-00077.fasta  scaffold_Scaffolds_0.cp.600  1387439   \n",
       "..                                  ...                          ...      ...   \n",
       "253  ./Inputfasta/SH0002528-00199.fasta                            1   328911   \n",
       "252  ./Inputfasta/SH0002528-00199.fasta                            1   326851   \n",
       "254  ./Inputfasta/SH0002528-00199.fasta                            1  2718930   \n",
       "255  ./Inputfasta/SH0002531-00147.fasta            scaffold_1.cp.600   839789   \n",
       "256  ./Inputfasta/SH0002531-00147.fasta            scaffold_1.cp.600  2253295   \n",
       "\n",
       "         END STRAND     GENE     COVERAGE     COVERAGE_MAP GAPS %COVERAGE  \\\n",
       "0    3404045      +   tet(O)  1-1920/1920  ===============  0/0       100   \n",
       "1      44353      +   mef(A)  1-1227/1227  ===============  0/0       100   \n",
       "2    3222364      -   tet(W)  1-1917/1923  ===============  0/0     99.69   \n",
       "3     671463      +   catA16   95-660/660  ..=============  0/0     85.76   \n",
       "4    1389355      -   tet(W)  1-1917/1923  ===============  0/0     99.69   \n",
       "..       ...    ...      ...          ...              ...  ...       ...   \n",
       "253   330131      +  tet(40)  1-1221/1221  ===============  0/0       100   \n",
       "252   328770      +   tet(O)  1-1920/1920  ===============  0/0       100   \n",
       "254  2719796      +     aadE    1-867/867  ===============  0/0       100   \n",
       "255   841708      +   tet(O)  1-1920/1920  ===============  0/0       100   \n",
       "256  2253789      +     dfrF    1-495/495  ===============  0/0       100   \n",
       "\n",
       "     ...  s_start    s_end evalue bit_score query_length subject_length  \\\n",
       "0    ...  11001.0      1.0    0.0   20116.0    3890762.0        11001.0   \n",
       "1    ...   9585.0   8157.0    0.0    2368.0    2324782.0        11179.0   \n",
       "2    ...      1.0   8027.0    0.0   14277.0    4962077.0         8027.0   \n",
       "3    ...      NaN      NaN    NaN       NaN          NaN            NaN   \n",
       "4    ...    193.0   2435.0    0.0    3938.0    2294929.0         8027.0   \n",
       "..   ...      ...      ...    ...       ...          ...            ...   \n",
       "253  ...   9442.0  10836.0    0.0    2499.0    3770667.0        10836.0   \n",
       "252  ...      1.0   2757.0    0.0    5092.0    3770667.0        11001.0   \n",
       "254  ...   1648.0      2.0    0.0    2863.0    3770667.0         8189.0   \n",
       "255  ...  10430.0      1.0    0.0   15830.0    3022989.0        10938.0   \n",
       "256  ...   1664.0    997.0    0.0    1223.0    3022989.0        15726.0   \n",
       "\n",
       "     perc_query_coverage_per_subject  subject sci names      ImmeDB_values  \\\n",
       "0                                0.0                NaN              IMEs1   \n",
       "1                                0.0                NaN  Genomic_islands50   \n",
       "2                                0.0                NaN             IMEs42   \n",
       "3                                NaN                NaN               None   \n",
       "4                                0.0                NaN             IMEs42   \n",
       "..                               ...                ...                ...   \n",
       "253                              0.0                NaN            IMEs393   \n",
       "252                              0.0                NaN              IMEs1   \n",
       "254                              0.0                NaN             IMEs35   \n",
       "255                              0.0                NaN              IMEs1   \n",
       "256                              1.0                NaN            IMEs383   \n",
       "\n",
       "                                All_immeDB_annotations  \n",
       "0             {'IMEs1': 20, 'ICEs78': 1, 'IMEs393': 1}  \n",
       "1                             {'Genomic_islands50': 1}  \n",
       "2                          {'IMEs42': 3, 'IMEs201': 1}  \n",
       "3                                                 None  \n",
       "4                          {'IMEs42': 3, 'IMEs201': 1}  \n",
       "..                                                 ...  \n",
       "253                       {'IMEs393': 1, 'IMEs460': 1}  \n",
       "252           {'IMEs1': 20, 'ICEs78': 1, 'IMEs393': 1}  \n",
       "254  {'IMEs35': 3, 'IMEs363': 1, 'IMEs369': 1, 'ICE...  \n",
       "255           {'IMEs1': 20, 'ICEs78': 1, 'IMEs393': 1}  \n",
       "256  {'IMEs383': 1, 'IMEs324': 1, 'IMEs419': 1, 'IM...  \n",
       "\n",
       "[257 rows x 32 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "abricate_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "abricate_df.to_csv(\n",
    "    \"/Users/sunitjain/Downloads/20230614_AMR-Abricate_report.extended_immedb.csv\",\n",
    "    index=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2.5kb extra for on each side."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aurum",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
